Source/WebCore/loader/SpeculationRules.cpp

```diff
@@ -0,0 +1,365 @@
+/*
+ * Copyright (C) 2025 Shopify Inc. All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#include "config.h"
+#include "SpeculationRules.h"
+
+#include <wtf/HashSet.h>
+#include <wtf/JSONValues.h>
+#include <wtf/URL.h>
+#include <wtf/URLHash.h>
+#include <wtf/Vector.h>
+#include <wtf/text/WTFString.h>
+
+namespace WebCore {
+
+Ref<SpeculationRules> SpeculationRules::create()
+{
+    return adoptRef(*new SpeculationRules);
+}
+
+const Vector<SpeculationRules::Rule>& SpeculationRules::prefetchRules() const
+{
+    return m_prefetchRules;
+}
+
+SpeculationRules::DocumentPredicate::DocumentPredicate(PredicateVariant&& value)
+    : m_value(WTFMove(value))
+{
+}
+
+const SpeculationRules::DocumentPredicate::PredicateVariant& SpeculationRules::DocumentPredicate::value() const
+{
+    return m_value;
+}
+
+static std::optional<Vector<String>> parseStringOrStringList(JSON::Object& object, const String& key)
+{
+    Vector<String> result;
+    auto value = object.getValue(key);
+    if (!value)
+        return Vector<String> { };
+
+    if (value->type() == JSON::Value::Type::String) {
+        String stringValue = value->asString();
+        if (!stringValue.isNull()) {
+            result.append(stringValue);
+            return result;
+        }
+    }
+
+    if (value->type() == JSON::Value::Type::Array) {
```

Comment:
asArray checks this redundantly.  I'd replace all this with just:
if (RefPtr arrayValue = value->asArray())